### DISCUSSION TITLE
How should I properly handle a remote kernel notebook directory from a JupyterServerProvider?

### DISCUSSION THREAD

**Discussion by andyscho** [microsoft/vscode-jupyter#15011] [Questions and Answers]
After switching from the [now-deprecated API](https://github.com/microsoft/vscode-jupyter/blob/4ca00b28cf3b890593e9c84d73b65e67f51a8061/src/api.deprecated.d.ts#L20) to [the new API](https://github.com/microsoft/vscode-jupyter/blob/4ca00b28cf3b890593e9c84d73b65e67f51a8061/src/api.d.ts#L27), I'm having issues handling the working directory for the notebook. Currently when I connect to a jupyter server in a JupyterServerCollection, it starts up a kernel with cwd as root filepath `/`. 

Given the current [createConnectionInfo](https://github.com/microsoft/vscode-jupyter/blob/main/src/kernels/jupyter/connection/jupyterConnection.ts#L49-L66) code, what is the recommended way to handle this? Should I be trying to do things through the proposed kernel API?

**Comment by DonJayamanne** [comment]
Sorry for not getting to this issue any sooner.
Unfortunately the Jupyter extension does not change the cwd of remote kernels.

> do things through the proposed kernel API?

If you are referring to the execution API, then yes thats an option.
I.e. you can run some code that will change the cwd.

Let me know if this works, and/or if you run into other issues.

**Reply by andyscho** [reply]
Thanks for the response.

Additionally, I'm curious about how best to shutdown remote kernels after switching to a different remote kernel (on the same server or a different server). There are some potentially helpful unstable IExportedKernelService APIs, but if there's a better supported way, I'd be interested.

**Reply by andyscho** [reply]
Also, I'd be curious if there's a way to disable being able to connect to a remote kernel that is running on a different notebook.

**Reply by DonJayamanne** [reply]
> Additionally, I'm curious about how best to shutdown remote kernels after switching to a different remote kernel (on the same server or a different server). There are some potentially helpful unstable IExportedKernelService APIs, but if there's a better supported way, I'd be interested.

Ideally users should shut these down.
Extensions should not shutdown kernels.

> disable being able to connect to a remote kernel that is running on a different notebook.

Unfortunately there's no way to do that.
The only way you can try to do that is to check the POST arguments passed to the kernel REST API and possibly throw some error based on some pre-processing logic (but this needs to be done in the Jupyter server rest API).
